src/System.Management.Automation/namespaces/FileSystemProvider.cs

```diff
@@ -31,6 +31,56 @@
 
 namespace Microsoft.PowerShell.Commands
 {
+    #region InodeTracker
+    /// <summary>
+    /// Tracks visited files/directories by caching their device IDs and inodes.
+    /// </summary>
+    internal class InodeTracker
+    {
+        private Dictionary<(UInt64, UInt64), bool>  _visitations;
```

Comment:
We use new Dictionary.TryAdd() but HashSet.Add() works the same - so we can use HashSet.